src/models/hf/checkpoint.rs

```diff
@@ -0,0 +1,149 @@
+///
+/// This trait is used to load a safetensors checkpoint from the repository.
+/// We just use it so that we can implement these methods on `Repo` rather
+/// than having them as standalone functions.
+trait LoadHFSafeTensors {
+    fn load_safetensors(&self) -> Result<Box<dyn SimpleBackend>, BoxedError>;
+
+    fn load_safetensors_multi(&self, index_path: &Path) -> Result<Vec<PathBuf>, BoxedError>;
+
+    fn load_safetensors_single(&self) -> Result<Vec<PathBuf>, BoxedError>;
+}
+
+impl<R> LoadHFSafeTensors for R
+where
+    R: Repo,
+{
+    /// Load a safetensors checkpoint.
+    ///
+    /// This method will first probe if there is a shard index. If there is,
+    /// a sharded checkpoint will be loaded. Otherwise, a single-file
+    /// checkpoint is loaded.
+    fn load_safetensors(&self) -> Result<Box<dyn SimpleBackend>, BoxedError> {
+        let file = self.file(SAFETENSORS_INDEX).context(DownloadSnafu {
+            name: SAFETENSORS_INDEX,
+        })?;
+
+        let paths = match file {
+            // We have a safetensor index, so load from shards.
+            Some(index_path) => self.load_safetensors_multi(&index_path),
+
+            // No index file, so assume that there is a single checkpoint.
+            None => self.load_safetensors_single(),
+        }?;
+
+        Ok(Box::new(unsafe {
+            MmapedSafetensors::multi(&paths).context(LoadCheckpointSnafu)?
+        }))
+    }
+
+    /// Load sharded safetensors checkpoint.
+    fn load_safetensors_multi(&self, index_path: &Path) -> Result<Vec<PathBuf>, BoxedError> {
+        // Parse the shard index.
+        let index_file = BufReader::new(
+            File::open(index_path).context(OpenSafeTensorsIndexSnafu { path: index_path })?,
+        );
+        let index: SafeTensorsIndex = serde_json::from_reader(index_file)
+            .context(ParseSafeTensorsIndexSnafu { path: index_path })?;
+
+        let mut shards = Vec::with_capacity(index.shards().len());
+        for shard_name in index.shards() {
+            let path = self.file(&shard_name).context(DownloadSnafu {
+                name: shard_name.clone(),
+            })?;
+            ensure!(path.is_some(), NonExistentShardSnafu { name: shard_name });
+            shards.push(path.unwrap());
+        }
+
+        Ok(shards)
+    }
+
+    /// Load non-sharded safetensors checkpoint.
+    fn load_safetensors_single(&self) -> Result<Vec<PathBuf>, BoxedError> {
+        let path = self.file(SAFETENSORS_SINGLE).context(DownloadSnafu {
+            name: SAFETENSORS_SINGLE,
+        })?;
+
+        ensure!(
+            path.is_some(),
+            NonExistentCheckpointSnafu {
+                name: SAFETENSORS_SINGLE.to_string(),
+            }
+        );
+
+        Ok(vec![path.unwrap()])
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct SafeTensorsIndex {
+    weight_map: HashMap<String, String>,
+}
+
+impl SafeTensorsIndex {
+    /// Get the names of the shards.
+    fn shards(&self) -> HashSet<String> {
```

Comment:
Can't we just borrow here?

Reply:
Notice that `SafeTensorsIndex` has a `HashMap`. I had the same reaction after coming back to this code after a week or so :).

Reply:
Ah, right. On the other hand, `SafeTensorsIndex` is a private struct, right? Why don't we directly iterate over the hash map instead? Is there an advantage to converting the values to a set first?

Reply:
The `HashMap` is: *parameter name -> shard filename*. So we cannot iterate over the hash map's values, because we need to remove duplicates first (multiple parameters in the same shard).